### PR TITLE
Add all version 5 options and support updating

### PR DESCRIPTION
**Pull request by edestecd**
See this for the same changes made to filebeat:
https://github.com/pcfens/puppet-filebeat/pull/113

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -12,41 +12,84 @@
...
+  #### v5 only ####
+  $use_generic_template = $winlogbeat::params::use_generic_template,
+  $beat_name            = $winlogbeat::params::beat_name,
+  $tags                 = $winlogbeat::params::tags,
+  $queue_size           = $winlogbeat::params::queue_size,
+  $max_procs            = $winlogbeat::params::max_procs,
+  $fields               = $winlogbeat::params::fields,
+  $fields_under_root    = $winlogbeat::params::fields_under_root,
+  $metrics              = undef,
+  #### End v5 onlly ####
```
Comment (claflico) [inline manifests/init.pp]: typo - "only"

File: manifests/params.pp [hunk 2]
```diff
@@ -1,30 +1,38 @@
...
+  # } else {
+  #   $conf_template = "${module_name}/winlogbeat.yml.erb"
+  # }
 
   case $::kernel {
     'Windows' : {
+      $package_ensure   = '5.1.1'
       $config_file      = 'C:/Program Files/Winlogbeat/winlogbeat.yml'
-      $download_url     = "https://download.elastic.co/beats/winlogbeat/winlogbeat-${package_version}-windows.zip"
+      $registry_file    = 'C:/ProgramData/winlogbeat/.winlogbeat.yml'
```
Comment (claflico) [inline manifests/params.pp]: Any particular reason for making this a different directory than the install_dir?

File: manifests/config.pp [hunk 3]
```diff
@@ -1,22 +1,32 @@
...
+    },
+    'output'            => $winlogbeat::outputs,
+    'shipper'           => $winlogbeat::shipper,
+    'logging'           => $winlogbeat::logging,
+    'runoptions'        => $winlogbeat::run_options,
+  })
 
   case $::kernel {
     'Windows' : {
+      $winlogbeat_path = 'c:\Program Files\Winlogbeat\winlogbeat.exe'
```
Comment (claflico) [inline manifests/config.pp]: Should this reference the install_dir variable to get the directory?

File: lib/facter/winlogbeat_version.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+require 'facter'
+Facter.add('winlogbeat_version') do
+  confine :kernel => %w(Windows) # rubocop:disable Style/HashSyntax
+  if File.exist?('c:\Program Files\Winlogbeat\winlogbeat.exe')
```
Comment (claflico) [inline lib/facter/winlogbeat_version.rb]: I haven't learned this part of puppet very well yet but can this file read in the install_dir as a variable?

File: manifests/params.pp (hunk 2 above)
Comment (edestecd) [inline manifests/params.pp]: So it is preserved on updates to a newer version. The update involves deleting C:/Program Files/Winlogbeat and replacing it with the contents of the new zip.  You want to keep the registry file so it does not resend all the data it already sent after the upgrade...

Also this is recommended from elastic: https://www.elastic.co/guide/en/beats/winlogbeat/current/configuration-winlogbeat-options.html#_registry_file

File: manifests/init.pp (hunk 1 above)
Comment (edestecd) [inline manifests/init.pp]: Copied from here: https://github.com/pcfens/puppet-filebeat/blob/master/manifests/init.pp#L86

I'll need to fix it there also...

File: lib/facter/winlogbeat_version.rb (hunk 4 above)
Comment (edestecd) [inline lib/facter/winlogbeat_version.rb]: Unfortunately no. Facts do not have access to manifests or their variables...
Facts are loaded first then sent to be used in the manifests.

File: manifests/params.pp (hunk 2 above)
Comment (claflico) [inline manifests/params.pp]: :thumbsup:

File: manifests/config.pp (hunk 3 above)
Comment (edestecd) [inline manifests/config.pp]: Probably. I copied from the filebeat module, so it should probably be updated there as well. I had that thought but was lazy.

File: manifests/init.pp (hunk 1 above)
Comment (edestecd) [inline manifests/init.pp]: fixed

File: manifests/config.pp (hunk 3 above)
Comment (edestecd) [inline manifests/config.pp]: fixed

File: lib/facter/winlogbeat_version.rb (hunk 4 above)
Comment (edestecd) [inline lib/facter/winlogbeat_version.rb]: no way to fix this.... Worst case the fact does not work and you can manually specify the major_version you want to get the right config file format.

**Comment by edestecd** [comment]
@claflico any other changes we need to get this merged?

**Comment by claflico** [comment]
Not that I see. I'm using the Master branch in our DEV environment (yeah, yeah I know) and want to make sure that when I merge it I have time to fix any issues that might show up and prevent puppet from running.

**Comment by edestecd** [comment]
You could just tag master and use that in your dev branch instead.
Or make a new branch called dev or something. Then master would be free for merges.

**Comment by claflico** [comment]
Hold my beer....  :)

**Pull request merged**